### PR TITLE
ccall: report static compile-time load issues correctly

### LINKED ISSUE #34061: precompile shouldn't try opening dylibs
Running the following:

```
❯ julia --project -J"/home/kc/Applications/julia-1.3.0/lib/julia/sys.so"  --output-o=sys.o \
 -e 'Base.reinit_stdio(); Base.init_depot_path(); Base.init_load_path(); using Plots'
```

error with 

```jl
ERROR: LoadError: error compiling gr_draw_marker: could not load library "libGR.so"
libGR.so: cannot open shared object file: No such file or directory
Stacktrace:
 [1] precompile at ./essentials.jl:463 [inlined]
 [2] _precompile_() at /home/kc/.julia/packages/Plots/2KhB2/src/precompile.jl:3
 [3] top-level scope at /home/kc/.julia/packages/Plots/2KhB2/src/Plots.jl:226
```

when Plots tries to execute one of the precompile statements in https://github.com/JuliaPlots/Plots.jl/blob/master/src/precompile.jl. The reason it can't find it is because with `--output-o` the `__init__` function doesn't run and `GR` makes its library availably to `Base.DL_LOAD_PATH` using that function.

It doesn't seem like `precompile` should go and try open dynamic libraries at least not in `--output-o` mode.

### PR DESCRIPTION
**Pull request by vtjnash**
fix #34061

Should have a test added that checks that something like:
```
./julia --output-o=x.o -e 'Base.reinit_stdio(); 
    f() = ccall((:dne, :does_not_exist), Cvoid, ());
    precompile(f, ())'
```
doesn't crash, or that calling `f()` exits the process early (no output-o file) with the message:
```
ERROR: could not load library "does_not_exist"
dlopen(does_not_exist.dylib, 1): image not found
Stacktrace:
 [1] f() at ./none:2
 [2] top-level scope at none:3
```

### CODE REVIEW

**Comment by KristofferC** [comment]
Confirmed this fixes the original use-case that cause med to file https://github.com/JuliaLang/julia/issues/34061.

**Comment by KristofferC** [comment]
Pushed a test

**Comment by JeffBezanson** [comment]
Ready to merge?

File: test/ccall.jl [hunk 1]
```diff
@@ -1535,3 +1535,11 @@ let
     @test isa(ptr, Ptr{Cvoid})
     @test arr[1] == '0'
 end
+
+# issue #34061
+sysimg = unsafe_string(Base.JLOptions().image_file)
+o_file = tempname()
+run(`$(Base.julia_cmd()) -J$sysimg --output-o=$o_file -e 'Base.reinit_stdio();
```
Comment (vtjnash) [inline test/ccall.jl]:
```suggestion
run(`$(Base.julia_cmd()) --output-o=$o_file -e 'Base.reinit_stdio();
```

**Review by vtjnash — COMMENTED** [review]
This test may be too slow (adds maybe 1 to 4 minutes) for something fairly trivial

File: test/ccall.jl (hunk 1 above)
Comment (KristofferC) [reply test/ccall.jl]: Oh yeah, that's already included in `julia_cmd`

**Comment by vtjnash** [comment]
@KristofferC You didn't address my review comment. Also, please cleanup up the commit messages when merging.

**Comment by KristofferC** [comment]
Strange, I completely missed that comment. I can change it to the fast fail version.

**Pull request merged**